salt/config/__init__.py

```diff
@@ -1144,6 +1144,30 @@ def _gather_buffer_space():
     # part of the extra_minion_data param
     # Subconfig entries can be specified by using the ':' notation (e.g. key:subkey)
     'pass_to_ext_pillars': (six.string_types, list),
+
+    # SSDP discovery publisher description.
+    # Contains publisher configuration and minion mapping.
+    # Setting it to False disables discovery
+    'discovery': (dict, bool),
+
+    # SSDP discovery mapping
+    # Defines arbitrary data for description and grouping minions across various types of masters,
+    # especially when masters are not related to each other.
+    'mapping': dict,
+
+    # SSDP discovery mapping matcher policy
+    # Values: "any" where at least one key/value pair should be found or
+    # "all", where every key/value should be identical
+    'match': str,
+
+    # Port definition.
+    'port': int,
```

Comment:
I am concerned that we are using the key of `port` here as a top-level key in this dictionary where it really only applies to the auto-discovery feature. I think we should prefix all of these new keys with `discovery_`

Reply:
Uhm... As I understand, this is just a key for whatever. So it is always like:
```yaml
discovery:
    port: 1234
```
What you suggesting is to have:
```yaml
discovery:
    discovery_port: 1234
```
I don't think this is any nice.

Reply:
@cachedout but if you thing that discovery is using `port` at the top level, then no way. I would never design something like that. :laughing: However, I thought that the `port` as a definition on its own can be happily used anywhere else by anything else in the future:
```yaml
some_new_stuff:
    port: 12345

pancake_burner:
    port: 12346

discover:  # as it is now
    port: 12347
```
...etc.

Reply:
I am thinking of this in the context of the `VALID_OPTS` dictionary only. I understand that in the _actual config_ that it is nested but it is not that way in `VALID_OPTS`.

Reply:
@cachedout OK, maybe I don't understand how to properly declare it then. As I understood, normally when it is defined:
```python
foo = (bar, baz)
```
So logically, `bar` and `baz` should be previously defined somewhere:
```python
bar = int
baz = boolean
```
Hence, valid construction is:
```yaml
foo = (123, True)
```

No?

Please just tell me then where should I place it the way that it doesn't get this ugly prefix, but stays just `port`? In this case seem `port` is _just_ generic port. And there is no `port` at the root level, so it supposed to be invalid, once defined, unless the whole config definition is just broken.

Reply:
@thatch45 could you please help here? Seem we stuck.

Reply:
@cachedout Ok, I looked at the `VALID_OPTS` dict once again. Does it mean that it doesn't even designed to represent a schema, but only a top-level? I thought that it kind of represents a structure of the config. But looking at this again, seems like I misunderstood. So you worry that:
```python
VALID_OPTS = {
    'port': int
}
```
...will bring up `port` at the root level of the config. Where I thought that if there will be _anywhere_ key e.g. `fridge` or `james_bond`, then the whole thing should at least `log.error('...` this that such key is invalid _anywhere_, no matter it is inside some structure or at the root level. And the same validation should also `log.error('...` if the key `port` is a string or float etc., regardless where it is placed.

Am I wrong?

Reply:
@cachedout that said, my reasoning was this:

- Define a key `port` in `VALID_OPTS` to make sure it is a) valid at all and b) it is `int` checked.
- Use it in my structure.

But if you afraid that somebody will start using `port` at the top level for something like `pancake_cooker_checker`, then yeah, this is wrong. But at this point I'd rather start thinking about the entire _schema_ of the Config, rather then just a one-level dictionary that defines what it defines: a valid options.

Reply:
@isbm You are not wrong. The schema was not designed for this case and I agree with you that is a problem. You'll see other places where this has been worked around by prefixing the intended key with its parent and an underscore.

We should absolutely fix this in a separate PR.

If you'd like to go ahead with this as-is, I can live with it though.

Reply:
Exactly. I would be happy doing this. For this matter I would say, there is another related problems: config is not really _validated_ before the Master/Minion starts. It only crashes on syntax error, but if somebody will add:

```yaml
foo: bar
```
This won't warning you something like "Option `foo` is unknown". But it gets even worse — nothing will happen if you do just this:
```yaml
ipv6: Take two
ret_port: "4506"
color: yes
```
At this point I see literally zero validation, and this should be really addressed properly. In a separate PR.

Reply:
@cachedout as of prefixes, they are ugly and messy. :wink: You don't say in English "pizza with pizzapepperoni" and "potato with potatopepperoni". You say "<object> with <another-object>". So why not just define "pepperoni" and serve with pizza or potato separately? Pizza can contain potato as well, BTW. This is the same reason why do we have 10 different boto_* modules and other module/function names that do almost the same thing, but called slightly differently. We had that discussion at SUSECon2017 about namespaces, remember? :wink: